### PR TITLE
[SEDONA-106] add st line from text

### PR DESCRIPTION
**Pull request by ngaur9**
## Is this PR related to a proposed Issue?
[SEDONA-106](https://issues.apache.org/jira/browse/SEDONA-106)

## What changes were proposed in this PR?
Added ST_LineFromText

## How was this patch tested?
Added unit tests for Scala, Python and Java. Tested the website using mkdocs.

## Did this PR include necessary documentation updates?
Yes. Added descriptions in the source code.

### CODE REVIEW

**Review by jiayuasu — CHANGES_REQUESTED** [review]
Do not add LINE geometry. Line is a LineString geometry. Use it instead.

**Comment by ngaur9** [comment]
I have incorporated the changes.

File: docs/archive/api/sql/GeoSparkSQL-Constructor.md [hunk 1]
```diff
@@ -155,6 +155,24 @@ FROM polygontable
 SELECT ST_PolygonFromText('-74.0428197,40.6867969,-74.0421975,40.6921336,-74.0508020,40.6912794,-74.0428197,40.6867969', ',') AS polygonshape
 ```
 
+## ST_LineFromText
```
Comment (jiayuasu) [inline docs/archive/api/sql/GeoSparkSQL-Constructor.md:158]: This doc is put in wrong place

File: sql/src/main/scala/org/apache/spark/sql/sedona_sql/expressions/Constructors.scala [hunk 2]
```diff
@@ -94,6 +94,37 @@ case class ST_PolygonFromText(inputExpressions: Seq[Expression])
   }
 }
 
+/**
+  * Return a line from a string. The string must be plain string and each coordinate must be separated by a delimiter.
+  *
+  * @param inputExpressions
+  */
+case class ST_LineFromText(inputExpressions: Seq[Expression])
```
Comment (jiayuasu) [inline sql/src/main/scala/org/apache/spark/sql/sedona_sql/expressions/Constructors.scala:102]: This implementation is not consistent with PostGIS: https://postgis.net/docs/ST_LineFromText.html

The input string must be WKT. The logic should simply re-use the ST_GeomFromWKT

File: sql/src/test/scala/org/apache/sedona/sql/predicateTestScala.scala [hunk 3]
```diff
@@ -140,6 +140,37 @@ class predicateTestScala extends TestBaseScala {
       assert(equaldf.count() == 0, s"Expected 0 value but got ${equaldf.count()}")
 
     }
+
+    it("Passed ST_Equals for ST_LineFromText and ST_Polygon") {
```
Comment (jiayuasu) [inline sql/src/test/scala/org/apache/sedona/sql/predicateTestScala.scala:144]: Please create a simple test case. Not sure why this has anything to do with polygons.

**Review by jiayuasu — CHANGES_REQUESTED** [review]

File: docs/api/sql/Constructor.md [hunk 4]
```diff
@@ -172,6 +172,24 @@ FROM linestringtable
 SELECT ST_LineStringFromText('-74.0428197,40.6867969,-74.0421975,40.6921336,-74.0508020,40.6912794', ',') AS linestringshape
 ```
 
+## ST_LineFromText
+
+Introduction: Construct a Line from Text, delimited by Delimiter
+
+Format: `ST_LineFromText (Text:string, Delimiter:char)`
+
+Since: `v1.0.0`
```
[the remark this answers is not in the document]
Comment (ngaur9) [reply docs/api/sql/Constructor.md]: Done.

File: docs/archive/api/sql/GeoSparkSQL-Constructor.md (hunk 1 above)
Comment (ngaur9) [reply docs/archive/api/sql/GeoSparkSQL-Constructor.md]: I have moved below ST_GeomFromWKT. Let me know if it has to be moved to some other place or file.

File: sql/src/main/scala/org/apache/spark/sql/sedona_sql/expressions/Constructors.scala (hunk 2 above)
Comment (ngaur9) [reply sql/src/main/scala/org/apache/spark/sql/sedona_sql/expressions/Constructors.scala:102]: I referred ST_GeomFromWKT  function call and added a LINE filter. Let me know if it needs modification or refactoring.

File: sql/src/test/scala/org/apache/sedona/sql/predicateTestScala.scala (hunk 3 above)
Comment (ngaur9) [reply sql/src/test/scala/org/apache/sedona/sql/predicateTestScala.scala]: Removed this test case and added a simple test case in constructorTestScala.scala file.

File: docs/archive/api/sql/GeoSparkSQL-Constructor.md [hunk 5]
```diff
@@ -19,6 +19,19 @@ FROM polygontable
 SELECT ST_GeomFromWKT('POINT(40.7128,-74.0060)') AS geometry
 ```
 
+## ST_LineFromText
+
+Introduction: Construct a Line from Wkt. Unlimited UUID strings can be appended.
+
```
Comment (jiayuasu) [inline docs/archive/api/sql/GeoSparkSQL-Constructor.md:25]: Again, do NOT change this doc please.

File: sql/src/main/scala/org/apache/spark/sql/sedona_sql/expressions/Constructors.scala [hunk 6]
```diff
@@ -94,6 +94,36 @@ case class ST_PolygonFromText(inputExpressions: Seq[Expression])
...
+  // This is an expression which takes one input expressions.
+  assert(inputExpressions.length == 1)
+
+  override def nullable: Boolean = false
+
+  override def eval(inputRow: InternalRow): Any = {
+    val lineString = inputExpressions(0).eval(inputRow).asInstanceOf[UTF8String].toString
+
+    var fileDataSplitter = FileDataSplitter.WKT
+    var formatMapper = new FormatMapper(fileDataSplitter, false, GeometryType.LINESTRING)
```
[the remark this answers is not in the document]
Comment (ngaur9) [reply sql/src/main/scala/org/apache/spark/sql/sedona_sql/expressions/Constructors.scala]: I just wanted to ask, if we remove the LineString, how will we identify if it is a LIne or some other geometrical figure?

[the remark this answers is not in the document]
Comment (jiayuasu) [reply sql/src/main/scala/org/apache/spark/sql/sedona_sql/expressions/Constructors.scala]: After you get the geomtetry, use geometry.getGeometryType to check if it is LineString. If not, return null. see https://locationtech.github.io/jts/javadoc/org/locationtech/jts/geom/Geometry.html#getGeometryType--

[the remark this answers is not in the document]
Comment (ngaur9) [reply sql/src/main/scala/org/apache/spark/sql/sedona_sql/expressions/Constructors.scala]: Added the Linestring filter to it. Please, let me know if anything else needs to be done.

File: sql/src/main/scala/org/apache/spark/sql/sedona_sql/expressions/Constructors.scala [hunk 7]
```diff
@@ -94,6 +94,40 @@ case class ST_PolygonFromText(inputExpressions: Seq[Expression])
...
+  * Return a line from a string. The string must be plain string and each coordinate must be separated by a delimiter.
+  *
+  * @param inputExpressions
+  */
+case class ST_LineFromText(inputExpressions: Seq[Expression])
+  extends Expression with CodegenFallback with UserDataGeneratator {
+  // This is an expression which takes one input expressions.
+  assert(inputExpressions.length == 1)
+
+  override def nullable: Boolean = false
```
Comment (jiayuasu) [inline sql/src/main/scala/org/apache/spark/sql/sedona_sql/expressions/Constructors.scala:107]: nullable should be true

File: docs/api/sql/Constructor.md [hunk 8]
```diff
@@ -172,6 +172,24 @@ FROM linestringtable
...
+
+Since: `v1.2.1`
+
+Spark SQL example:
+```SQL
+SELECT ST_LineFromText(linetable._c0,',') AS lineshape
+FROM linetable
+```
+
+```SQL
```
Comment (jiayuasu) [inline docs/api/sql/Constructor.md:189]: This example should use WKT string as input. Please fix the doc accordingly

**Pull request merged**